src/ui/popup.js

```diff
@@ -512,6 +522,47 @@ export default class Popup extends Evented {
         this._update(event.point);
     }
 
+    _getAnchor(offset: any) {
+        if (this.options.anchor) { return this.options.anchor; }
+
+        const pos: any = this._pos;
+        const width = this._container.offsetWidth;
+        const height = this._container.offsetHeight;
+
+        let anchorComponents;
+        if (pos.y + offset.bottom.y < height) {
+            anchorComponents = ['top'];
+        } else if (pos.y > this._map.transform.height - height) {
+            anchorComponents = ['bottom'];
+        } else {
+            anchorComponents = [];
+        }
+
+        if (pos.x < width / 2) {
+            anchorComponents.push('left');
+        } else if (pos.x > this._map.transform.width - width / 2) {
+            anchorComponents.push('right');
+        }
+
+        if (anchorComponents.length === 0) {
+            return 'bottom';
+        }
+        return ((anchorComponents.join('-'): any): Anchor);
+
+    }
+
+    _updateClassList() {
+        const classes = new Set(this._classList);
+        classes.add('mapboxgl-popup');
+        if (this._anchor) {
+            classes.add(`mapboxgl-popup-anchor-${this._anchor}`);
+        }
+        if (this._trackPointer) {
+            classes.add('mapboxgl-popup-track-pointer');
+        }
+        this._container.className = [...classes].join(' ');
```

Comment:
I think this could be rearranged to reduce a set allocation (spread to arrray, then `.push` optional classes, then join), but it's so far down in the list of memory usage, that I fully support the above 👍

Reply:
Good idea!